text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+from the main thread would go to, which is normally "standard error".
+(An application-controllable override mechanism may make sense,
+[see below][squelching-diagnostics].)
+
+The `progname` argument is a short version of the program's name
+(abstractly, what you would type at the command line to start the
+program, assuming it were in `PATH`; concretely, the basename of
+`argv[0]`, with any trailing `.exe` or `.com` chopped off on Windows,
+but not on other platforms).  This makes it easy to generate
+diagnostics in the style conventional for "Unixy" systems, e.g.
+
+    grep: /etc/shadow: Permission denied
+
+`exit_status` shall convert `self` to a platform-specific exit code,
+conveying at least a notion of success or failure.  The return type is
+`i32` to match [std::process::exit][] (which probably calls the C
+library's `exit` primitive), but (as already documented for
+`process::exit`) on "most Unix-like" operating systems, only the low 8
+bits of this value are significant.
+
+[std::process::exit]: https://doc.rust-lang.org/std/process/fn.exit.html
+
+## Changes to `main`
+[changes-to-main]: #changes-to-main
+
+From the perspective of the code that calls `main`, its signature is
+now generic:
+
+``` rust
+fn<T: Termination> main() -> T { ... }
+```
+
+It is critical that people _writing_ main should not have to treat it
+as a generic, though.  Existing code with `fn main() -> ()` should
+continue to compile, and code that wants to use the new feature should
+be able to write `fn main() -> TermT` for some concrete type `TermT`.
+
+I also don't know whether the code that calls `main` can accept a
+generic.  The `lang_start` glue currently receives an unsafe pointer
+to `main`, and expects it to have the signature `() -> ()`.  The
+abstractly correct new signature is `() -> Termination` but I suspect
+that this is currently not possible, pending `impl Trait` return types
+or something similar.  `() -> Box<Termination>` probably _is_ possible
+but requires a heap allocation, which is undesirable in no-std
+contexts.
+
+A solution to both problems is to implement the new semantics entirely
+within the compiler.  When it notices that `main` returns anything
+other than `()`, it renames the function and injects a shim:
+
+``` rust
+fn main() -> () {
+    let term = $real_main();
+    term.write_diagnostic(std::env::progname(), std::io::LOCAL_STDERR);
+    std::process::exit(term.exit_status());
+}
+```
+
+## `main` in nostd environments
+[main-in-nostd-environments]: #main-in-nostd-environments
+
+Some no-std environments do have a notion of processes that run and
+then exit, but they may or may not have notions of "exit status" or
+"error messages".  In this case, the signature of `main` should be
+unchanged, and the shim should simply ignore whichever aspects of
+`Termination` don't make sense in context.
+
+There are also environments where
+[returning from `main` constitutes a _bug_.][divergent-main] If you
```

Comment:
I'm not at all a microcontroller person.  Divergent `main` is in this RFC because it was [explicitly asked for over here](https://internals.rust-lang.org/t/allowing-for-main-to-be-divergent-in-embedded-environments/4717).  Are you saying that that's not actually a useful feature?